stack-graphs/src/serde.rs

```diff
@@ -0,0 +1,643 @@
+use crate::{arena::Handle, json::Filter};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Default)]
+pub struct StackGraph {
+    files: Files,
+    nodes: Nodes,
+    edges: Edges,
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum Error {
+    #[error("failed to load file `{0}`")]
+    FileNotFound(String),
+
+    #[error("duplicate file `{0}`")]
+    FileAlreadyPresent(String),
+
+    #[error("failed to locate node `{0}` in graph")]
+    NodeNotFound(u32),
+
+    #[error("no file data for node `{0}`")]
+    NoFileData(u32),
+}
+
+impl StackGraph {
+    pub fn from_graph<'a>(graph: &crate::graph::StackGraph, filter: &'a dyn Filter) -> Self {
```

Comment:
Similarly, we could call this `from_graph_filter` and have a simply `from_graph` implementation that passes the `NoFilter` argument.